idaplugin/rematch/actions/match.py

```diff
@@ -1,63 +1,178 @@
+    except StopIteration:
+      self.timer.stop()
+      return
 
-      func = instances.FunctionInstance(self.file_id, offset)
-      network.query("POST", "collab/instances/", params=func.serialize(),
-                    json=True)
+    try:
+      func = instances.FunctionInstance(netnode.bound_file_id, offset)
+      self.instance_set.append(func.serialize())
 
-      i = i + 1
-      self.progress.setValue(i)
-      if (i >= self.progress.maximum()):
-        self.timer.stop()
-    except:
-      self.timer.stop()
+      if len(self.instance_set) >= 100:
+        network.delayed_query("POST", "collab/instances/",
+                              params=self.instance_set, json=True,
+                              callback=self.progress_advance)
+        self.instance_set = []
+        self.pbar.setMaximum(self.pbar.maximum() + 1)
+      self.progress_advance()
+    except Exception:
+      self.cancel_upload()
       raise
 
-  def cancel(self):
+  def progress_advance(self, result=None):
+    del result
+    new_value = self.pbar.value() + 1
+    self.pbar.setValue(new_value)
+    if new_value >= self.pbar.maximum():
+      self.pbar.accept()
+
+  def cancel_upload(self):
     self.timer.stop()
+    self.timer = None
+    self.pbar = None
 
+  def reject_upload(self):
+    self.cancel_upload()
 
-class MatchFunctionAction(base.BoundFileAction):
-  name = "Match &Function"
-  group = "Match"
+  def accept_upload(self):
+    self.cancel_upload()
 
-  @staticmethod
-  def activate(ctx):
-    file_id = netnode.bound_file_id
+    if self.source == 'idb':
+      self.source_range = [None, None]
+    elif self.source == 'single':
+      self.source_range = [self.source_single, self.source_single]
+    elif self.source == 'range':
+      pass
+    else:
+      raise NotImplementedError("Unsupported source type encountered in task "
+                                "creation")
 
-    function = idaapi.choose_func("Choose function to match with database",
-                                  idc.ScreenEA())
-    if function is None:
-      return
+    params = {'source_file': netnode.bound_file_id,
+              'source_start': self.source_range[0],
+              'source_end': self.source_range[1],
+              'target_project': self.target_project,
+              'target_file': self.target_file,
+              'source': self.source, 'methods': self.methods}
+    r = network.query("POST", "collab/tasks/", params=params, json=True)
+    self.task_id = r['id']
+
+    self.pbar = QtWidgets.QProgressDialog()
+    self.pbar.setLabelText("Waiting for remote matching... You may continue "
+                           "working without any limitations.")
+    self.pbar.setRange(0, int(r['progress_max']) if r['progress_max'] else 0)
+    self.pbar.setValue(int(r['progress']))
+    self.pbar.canceled.connect(self.cancel_task)
+    self.pbar.rejected.connect(self.reject_task)
+    self.pbar.accepted.connect(self.accept_task)
+    self.pbar.show()
```

Comment:
It's more of a cosmetic thing, rather than having all this code written there, create a class, move it to the proper place, connect it later